node/consensus/data/message_validators.go

```diff
@@ -0,0 +1,99 @@
+package data
+
+import (
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+	"source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb"
+	"source.quilibrium.com/quilibrium/monorepo/node/protobufs"
+)
+
+func (e *DataClockConsensusEngine) validateFrameMessage(peerID peer.ID, message *pb.Message) int {
+	msg := &protobufs.Message{}
+	if err := proto.Unmarshal(message.Data, msg); err != nil {
+		e.logger.Debug("could not unmarshal message", zap.Error(err))
+		return -1
+	}
+	a := &anypb.Any{}
+	if err := proto.Unmarshal(msg.Payload, a); err != nil {
+		e.logger.Debug("could not unmarshal payload", zap.Error(err))
+		return -1
+	}
+	switch a.TypeUrl {
+	case protobufs.ClockFrameType:
+		frame := &protobufs.ClockFrame{}
+		if err := proto.Unmarshal(a.Value, frame); err != nil {
+			e.logger.Debug("could not unmarshal frame", zap.Error(err))
+			return -1
+		}
+		if ts := time.UnixMilli(frame.Timestamp); time.Since(ts) > time.Hour {
+			e.logger.Debug("frame is too old", zap.Time("timestamp", ts))
+			return 1
+		}
+		return 0
+	default:
+		e.logger.Debug("unknown message type", zap.String("type_url", a.TypeUrl))
+		return -1
+	}
+}
+
+func (e *DataClockConsensusEngine) validateTxMessage(peerID peer.ID, message *pb.Message) int {
+	msg := &protobufs.Message{}
+	if err := proto.Unmarshal(message.Data, msg); err != nil {
+		e.logger.Debug("could not unmarshal message", zap.Error(err))
+		return -1
+	}
+	a := &anypb.Any{}
+	if err := proto.Unmarshal(msg.Payload, a); err != nil {
+		e.logger.Debug("could not unmarshal payload", zap.Error(err))
+		return -1
+	}
+	switch a.TypeUrl {
+	case protobufs.TokenRequestType:
+		tx := &protobufs.TokenRequest{}
+		if err := proto.Unmarshal(a.Value, tx); err != nil {
+			e.logger.Debug("could not unmarshal token request", zap.Error(err))
+			return -1
+		}
+		return 0
```

Comment:
There is no timestamp associated with a token request. I am not sure if there should be one, but just FYI.